pkg/node/node.go

```diff
@@ -4,243 +4,203 @@
 	outType reflect.Type
 }
 
-func (s *Init) SendsTo(outputs ...Receiver) {
-	assertChannelsCompatibility(s.fun.ArgChannelType(0), outputs)
+func (s *Init[OUT]) SendsTo(outputs ...Receiver[OUT]) {
+	//assertChannelsCompatibility(s.fun.ArgChannelType(0), outputs)
 	s.outs = append(s.outs, outputs...)
 }
 
-func (s *Init) OutType() reflect.Type {
+func (s *Init[OUT]) OutType() reflect.Type {
 	return s.outType
 }
 
 // Middle is any intermediate node that receives data from another node, processes/filters it,
 // and forwards the data to another node.
 // An Middle node must have at least one output node.
-type Middle struct {
-	outs    []Receiver
-	inputs  connect.Joiner
+type Middle[IN, OUT any] struct {
+	outs    []Receiver[OUT]
+	inputs  connect.Joiner[IN]
 	started bool
-	fun     refl.Function
+	fun     MiddleFunc[IN, OUT]
 	outType reflect.Type
 	inType  reflect.Type
 }
 
-func (i *Middle) joiner() *connect.Joiner {
+func (i *Middle[IN, OUT]) joiner() *connect.Joiner[IN] {
 	return &i.inputs
 }
 
-func (i *Middle) isStarted() bool {
+func (i *Middle[IN, OUT]) isStarted() bool {
 	return i.started
 }
 
-func (s *Middle) SendsTo(outputs ...Receiver) {
-	assertChannelsCompatibility(s.fun.ArgChannelType(1), outputs)
+func (s *Middle[IN, OUT]) SendsTo(outputs ...Receiver[OUT]) {
 	s.outs = append(s.outs, outputs...)
 }
 
-func (m *Middle) OutType() reflect.Type {
+func (m *Middle[IN, OUT]) OutType() reflect.Type {
 	return m.outType
 }
 
-func (m *Middle) InType() reflect.Type {
+func (m *Middle[IN, OUT]) InType() reflect.Type {
 	return m.inType
 }
 
 // Terminal is any node that receives data from another node and does not forward it to another node,
 // but can process it and send the results to outside the graph (e.g. memory, storage, web...)
-type Terminal struct {
-	inputs  connect.Joiner
+type Terminal[IN any] struct {
+	inputs  connect.Joiner[IN]
 	started bool
-	fun     refl.Function
+	fun     TerminalFunc[IN]
 	done    chan struct{}
 	inType  reflect.Type
 }
 
-func (i *Terminal) joiner() *connect.Joiner {
+func (i *Terminal[IN]) joiner() *connect.Joiner[IN] {
 	return &i.inputs
 }
 
-func (t *Terminal) isStarted() bool {
+func (t *Terminal[IN]) isStarted() bool {
 	return t.started
 }
 
 // Done returns a channel that is closed when the Terminal node has ended its processing. This
 // is, when all its inputs have been also closed. Waiting for all the Terminal nodes to finish
 // allows blocking the execution until all the data in the graph has been processed and all the
 // previous stages have ended
-func (t *Terminal) Done() <-chan struct{} {
+func (t *Terminal[IN]) Done() <-chan struct{} {
 	return t.done
 }
 
-func (m *Terminal) InType() reflect.Type {
+func (m *Terminal[IN]) InType() reflect.Type {
 	return m.inType
 }
 
 // AsInit wraps an InitFunc into an Init node. It panics if the InitFunc does not follow the
 // func(chan<-) signature.
-func AsInit(fun InitFunc) *Init {
-	fn := refl.WrapFunction(fun)
```

Comment:
Removed reflection package, since we have changed runtime checks by compile-time checks!